### PR TITLE
Small neon cli to run neon components locally

### PR DESCRIPTION
**Pull request by NanoBjorn**
## Describe your changes

I have been playing with neon locally [with docker](https://www.percona.com/blog/using-docker-to-deploy-neon-serverless-postgresql/), tried `neon_local`, but it seems more for testing purposes and not allows enough agility (like running two pageservers or smth). Decided to write a POC for some single-entrypoint approach, which allows to run each component with single command without hidden configs, with explicit network and binary dependencies. So here it is, the readme is pretty straighforward.

TLDR how to run the components with the script:

```bash
# Create data directories for safekeeper, pageserver and compute
mkdir -p tmp/sk1 tmp/ps tmp/compute

# Run opentelemetry server, because without it compute will not start
docker run -p 4318:4318 --rm otel/opentelemetry-collector:latest

# Export these in all terminals, or use direnv
export OTEL_EXPORTER_OTLP_ENDPOINT=http://localhost:4318
export NEON_BINARIES_DIR="$(cd .. && pwd)/target/debug"

./neon broker

./neon safekeeper -D tmp/sk1 --id 1

./neon pageserver -D tmp/ps \
  --broker-http='127.0.0.1:50051' \
  --id=123 \
  -c "pg_distrib_dir='$(cd .. && pwd)/pg_install'"

# Register new tenant with timeline in pageserver and generate spec for compute 
./neon compute init \
  --pageserver-pg=127.0.0.1:6400 \
  --pageserver-http 127.0.0.1:9898 \
  --safekeeper-pg=127.0.0.1:5454 \
  --pgbin=$(cd .. && pwd)/pg_install/v15/bin/postgres \
  --spec-path tmp/compute_spec.json

# Start compute with given spec
./neon compute start \
  --pgdata tmp/compute/pgdata \
  --pgbin=$(cd .. && pwd)/pg_install/v15/bin/postgres \
  --spec-path tmp/compute_spec.json
```

There are arguments specified in help, all other arguments 
passed after these are passed to the binary of the component.
To show help of each binary, you can use `./neon help <component>`.
To show help of this script, use `./neon [SUBCOMMAND ...] --help`, 
like `./neon compute init --help`. 

## Checklist before requesting a review

- [x] I have performed a self-review of my code.
- [ ] ~~If it is a core feature, I have added thorough tests.~~
- [ ] ~~Do we need to implement analytics? if so did you add the relevant metrics to the dashboard?~~
- [ ] ~~If this PR requires public announcement, mark it with /release-notes label and add several sentences in this section.~~

## Checklist before merging

- [ ] Do not forget to reformat commit message to not include the above checklist

### CODE REVIEW

File: neon_cli/neon [hunk 1]
```diff
@@ -0,0 +1,326 @@
...
+
+BIN_DIR = os.getenv('NEON_BINARIES_DIR', './')
+
+
+def eprint(*args, **kwargs):
+    print(*args, file=sys.stderr, **kwargs)
+
+
+if sys.version_info.major != 3 and sys.version_info.minor < 9:
+    eprint('Warning: This script was written on python 3.9, your version is ')
```
Comment (vadim2404) [inline neon_cli/neon:26]: it seems that you forgot to add info about user's version

File: neon_cli/neon [hunk 2]
```diff
@@ -0,0 +1,326 @@
...
+
+# enable debug logging
+logging.basicConfig()
+logger = logging.getLogger()
+logger.setLevel(logging.DEBUG)
+
+BIN_DIR = os.getenv('NEON_BINARIES_DIR', './')
+
+
+def eprint(*args, **kwargs):
```
Comment (vadim2404) [inline neon_cli/neon:21]: but you can just use `logger.error`?

File: neon_cli/neon [hunk 3]
```diff
@@ -0,0 +1,326 @@
...
+        }
+    )
+    assert resp.status_code == 201
+    assert tenant_id == resp.json()['tenant_id']
+    timeline_id = resp.json()['timeline_id']
+    spec = COMPUTE_SPEC_TEMPLATE.replace('<TIMELINE_ID>', timeline_id)
+    spec = spec.replace('<SAFEKEEPERS>', ','.join(safekeerpers))
+    spec = spec.replace('<TENANT_ID>', tenant_id)
+
+    ps_host = pageserver_pg.split(':')[0]
```
Comment (vadim2404) [inline neon_cli/neon:80]:
```python
pg_host, pg_port = pageserver_pg.split(':', 1)
```

Comment (NanoBjorn) [reply neon_cli/neon]: fixed

File: neon_cli/neon (hunk 2 above)
Comment (NanoBjorn) [reply neon_cli/neon]: that's right, removed it

File: neon_cli/neon (hunk 1 above)
Comment (NanoBjorn) [reply neon_cli/neon]: added

**Comment by kelvich** [comment]
Few thoughts:
* I think extending `neon_local` is a proper way to integrate. @hlinnaka recently switched test to use more of `neon_local` instead of duplicating the same code in python. IIRC we have some tests with two pageservers, so that would be helpful to have in `neon_local`
* Somewhat an alternative approach is to have one binary with everything in it (e.g. neon_storage) and config file command what threads to start. So deployed pageserver can start only pageserver-related threads, but for the local setup we can start pageserver+safekeepers+broker as one binary.
* Once we publish control plane I think we can think of using control plane binary instead of the `neon_local`

**Comment by kelvich** [comment]
And I like the proposed command structure. Current one is more or less randomly evolved from the several ideas that we initially had about locally installed neon ([003-laptop-cli](https://github.com/neondatabase/neon/blob/main/docs/rfcs/003-laptop-cli.md), [006-laptop-cli-v2-CLI](https://github.com/neondatabase/neon/blob/main/docs/rfcs/006-laptop-cli-v2-CLI.md), [006-laptop-cli-v2-repository-structure](https://github.com/neondatabase/neon/blob/main/docs/rfcs/006-laptop-cli-v2-repository-structure.md), [008-push-pull](https://github.com/neondatabase/neon/blob/main/docs/rfcs/008-push-pull.md) -- most of them are not relevant now)

**Comment by NanoBjorn** [comment]
> at least it's used constantly through the test suite

> extending neon_local is a proper way to integrate

exactly, consider this piece of python code more as a prototype

> think of using control plane binary instead of the `neon_local`
this would be a bold move, I suppose control plane will be much more complicated, than something like `neon_local` or this tool

> config file command what threads to start
I was also thinking about it, more of docker-compose-like approach, where we will have all needed components with their configuration in a single config, but at our current point it would require much more work to be done, because this would introduce just another nested level (this new all-in-one config). Btw kubernetes supports both — you can create resources with `kubectl` and with `kubectl apply -f`, but from simplicity point of view in our case I still prefer cli arguments

As for extending `neon_local`, I can think about it (would be nice point to pick up rust 😄), I am not sure I am happy about naming, ideally all-in-one neon binary should be named as simple as possible, but I can start with extending neon_local to be able to run components independently, as in this tool

**Comment by NanoBjorn** [comment]
@kelvich @vadim2404 any ideas on what to do next with this pr? Merge it or I can move it to my personal repo just to let it exist and maybe will extend neon_local from time to time?

**Comment by NanoBjorn** [comment]
It was fun for a hackathon, but I don't think it is needed to be open anymore.

**Pull request closed without merging**